src/mesh/Draco_Mesh.cc

```diff
@@ -0,0 +1,265 @@
+    std::vector<std::vector<unsigned>> vec_node_vec(num_pairs,
+                                                    std::vector<unsigned>(2));
+    // TODO: change type of vec_node_vec?
+    unsigned k = 0;
+    for (unsigned i = 0; i < nper_cell; ++i) {
+      for (unsigned j = i + 1; j < nper_cell; ++j) {
+
+        // set kth pair entry to the size 2 vector of nodes
+        vec_node_vec[k] = {cell_to_node_map[cell][i],
+                           cell_to_node_map[cell][j]};
+
+        // increment k
+        k++;
+      }
+    }
+
+    Check(k == num_pairs);
+
+    // check if each pair constitutes a face
+    for (unsigned l = 0; l < num_pairs; ++l) {
+
+      // get adjacent cells from node-to-cell map
+      // TODO: add DbC to ensure these are sorted from step 2
+      const std::vector<unsigned> &vert0_cells =
+          node_to_cell_map[vec_node_vec[l][0]];
+      const std::vector<unsigned> &vert1_cells =
+          node_to_cell_map[vec_node_vec[l][1]];
+
+      // find common cells (set_intersection is low-complexity)
+      // TODO: reserve size for cells_in_common
+      std::vector<unsigned> cells_in_common;
+      std::set_intersection(vert0_cells.begin(), vert0_cells.end(),
+                            vert1_cells.begin(), vert1_cells.end(),
+                            std::back_inserter(cells_in_common));
+
+      // these nodes should have at least cell index "cell" in common
+      Check(cells_in_common.size() >= 1);
+
+      // populate cell-to-cell linkage
+      // TODO: populate face-to-cell (and inverse) map here?
+      if (cells_in_common.size() > 1) {
+        for (auto oth_cell : cells_in_common) {
+          if (oth_cell != cell)
+            cell_to_cell_linkage[cell].push_back(
+                std::make_pair(oth_cell, vec_node_vec[l]));
+        }
+      }
+
+      // check if this cell pair has a side flag
+      // TODO: add DbC to ensure these are sorted from step 3
+      const std::vector<unsigned> &vert0_sides =
+          node_to_side_map[vec_node_vec[l][0]];
+      const std::vector<unsigned> &vert1_sides =
+          node_to_side_map[vec_node_vec[l][1]];
+
+      // find common sides
+      std::vector<unsigned> sides_in_common;
+      std::set_intersection(vert0_sides.begin(), vert0_sides.end(),
+                            vert1_sides.begin(), vert1_sides.end(),
+                            std::back_inserter(sides_in_common));
+
+      Check(sides_in_common.size() <= 1);
+      if (sides_in_common.size() > 0) {
+        // populate cell-to-side linkage
+        // TODO: replace with typedef Draco_Layout::Boundary_Layout
+        cell_to_side_linkage[cell].push_back(
+            std::make_pair(sides_in_common[0], vec_node_vec[l]));
+      }
+
+      // TODO: add check for ghost face
+    }
+  }
+
+  // STEP 5: instantiate the full layout
+  // TODO: finish Draco_Layout class
+
+  Ensure(cell_to_cell_linkage.size() == num_cells);
+  Ensure(cell_to_side_linkage.size() == num_cells);
+}
```

Comment:
At some point, we will need to profile the performance of this routine for large meshes.  But for now, it is fine.

Reply:
This is a good idea.  Incidentally, there are parts of this routine that may be straightforward to thread (node-pair-wise face id'ing, for instance).